File: javascript/spec/set-spec.js
var Set = require("../set.js");

describe("Given an empty set when adding two new objects", function () {
	var set;

	beforeEach(function() {
		set = new Set();
		set.add("hello");
		set.add("world cup");
	});

	it("should hold the new objects in the set", function () {
		expect(set.indexOf("hello")).toBe(0);
		expect(set.indexOf("world cup")).toBe(1);
	});

	it("should increment the count to two", function () {
		expect(set.count()).toBe(2);
	});

	it("should increment the count to three when adding a further object", function () {
		set.add("alright");
		expect(set.count()).toBe(3);
	});
});

describe("Given an set with three objects", function () {
	var set;

	beforeEach(function () {
		set = new Set();
		set.add(1);
		set.add(120);
		set.add(453);
	});

	it("should throw an exception if object is not in set", function () {
		expect(function () { set.indexOf(88) }).toThrow();
	});

	it("should be able to match more than one object doing a count", function () {
		expect(set.count(function (value) { return value > 2 })).toBe(2);
	});

	it("should return zero if count criteria is not met", function () {
		expect(set.count(function (value) { return value < 1 })).toBe(0);
	});
});

File: ruby/spec/set_spec.rb
require 'myset'

RSpec.describe MySet do
	describe 'add' do
		before(:each) do
			@set = MySet.new
		end

		it 'increments the count' do
			@set.add("hello")
			expect(@set.count()).to eq(1)
			@set.add("agh")
			expect(@set.count()).to eq(2)
		end

		it 'holds two added objects' do
			@set.add("hello")
			@set.add("agh")
			expect(@set.indexOf("agh")).to eq(1)
			expect(@set.indexOf("hello")).to eq(0)
		end
	end

	describe 'indexOf' do
		it 'throws an exception if it does not contain object' do
			set = MySet.new
			set.add(44)
			set.add(102)
			expect { set.indexOf(55) }.to raise_error
		end
	end

	describe 'any' do
		before(:each) do
			@set = MySet.new
			@set.add(44)
			@set.add(102)
			@set.add(1022)
		end

		it 'returns true when any object is found' do
			expect(@set.any(method(:greaterThan55))).to eq(true)
		end

		it 'returns false when any object is not found' do
			expect(@set.any(method(:lessThan44))).to eq(false)
		end
	end

	def greaterThan55(value)
		return value > 55
	end

	def lessThan44(value)
		puts value
		return value < 44
	end
end